packages/eslint-plugin/src/rules/no-unused-vars.ts

```diff
@@ -557,40 +560,71 @@ export default createRule<Options, MessageIds>({
     }
 
     return {
-      // declaration file handling
-      [ambientDeclarationSelector(AST_NODE_TYPES.Program, true)](
+      // top-level declaration file handling
+      [ambientDeclarationSelector(AST_NODE_TYPES.Program)](
         node: DeclarationSelectorNode,
       ): void {
         if (!isDefinitionFile(context.filename)) {
           return;
         }
+
+        const moduleDecl = nullThrows(
+          node.parent,
+          NullThrowsReasons.MissingParent,
+        ) as TSESTree.Program;
+
+        if (checkForOverridingExportStatements(moduleDecl)) {
+          return;
+        }
+
         markDeclarationChildAsUsed(node);
       },
 
       // children of a namespace that is a child of a declared namespace are auto-exported
       [ambientDeclarationSelector(
         'TSModuleDeclaration[declare = true] > TSModuleBlock TSModuleDeclaration > TSModuleBlock',
-        false,
       )](node: DeclarationSelectorNode): void {
+        const moduleDecl = nullThrows(
+          node.parent.parent,
+          NullThrowsReasons.MissingParent,
+        ) as TSESTree.TSModuleDeclaration;
+
+        if (checkForOverridingExportStatements(moduleDecl)) {
+          return;
+        }
```

Comment:
This is necessary so the following is reported ([playground link](https://typescript-eslint.io/play/#ts=5.7.3&fileType=.ts&code=KYDwDg9gTgLgBAE2AYwDYEMrDgO3QW2AGcx1lsAxCCOAbwCg5cDjTy4AhTOxp5wkmWxcAXjz58A9JLhEAFhACuqBHABG2LJFjBVaxfACWRHAHIYvCXgFthBcRL7IIOIvBBwAvHACMAbks%2BAF96QKZQbXhaOCCAiRCmEKCgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)):

```ts
export declare namespace Foo {
  namespace Bar {
    namespace Baz {
      // should be reported but isn't
      namespace Bam {
        const x = 1;
      }

      export { };
    }
  }
}
```